File: java/note-taking/src/test/java/com/rbleggi/notetaking/NoteManagerTest.java
package com.rbleggi.notetaking;

import org.junit.jupiter.api.*;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

class NoteManagerTest {
    private NoteManager manager;
    private static final String TEST_FILE = "test_notes.txt";

    @BeforeEach
    void setUp() {
        manager = new NoteManager();
    }

    @AfterEach
    void tearDown() throws IOException {
        Files.deleteIfExists(Path.of(TEST_FILE));
    }

    @Test
    @DisplayName("AddNoteCommand should add note to manager")
    void addNoteCommand_addsNote() throws Exception {
        var command = new AddNoteCommand(manager, "Title", "Content");
        command.execute();

        var notes = manager.listNotes();
        assertEquals(1, notes.size());
        assertEquals("Title", notes.get(0).title());
        assertEquals("Content", notes.get(0).content());
    }

    @Test
    @DisplayName("NoteManager should add multiple notes with unique IDs")
    void noteManager_addMultipleNotes_assignsUniqueIds() {
        var note1 = manager.addNote("First", "Content 1");
        var note2 = manager.addNote("Second", "Content 2");

        assertNotEquals(note1.id(), note2.id());
        assertEquals(2, manager.listNotes().size());
    }

    @Test
    @DisplayName("NoteManager should edit existing note")
    void noteManager_editNote_updatesContent() {
        var note = manager.addNote("Original", "Original Content");
        var edited = manager.editNote(note.id(), Optional.of("Updated"), Optional.empty());

        assertTrue(edited.isPresent());
        assertEquals("Updated", edited.get().title());
        assertEquals("Original Content", edited.get().content());
    }

    @Test
    @DisplayName("NoteManager should delete note")
    void noteManager_deleteNote_removesNote() {
        var note = manager.addNote("To Delete", "Content");
        assertTrue(manager.deleteNote(note.id()));
        assertEquals(0, manager.listNotes().size());
    }

    @Test
    @DisplayName("NoteManager should return false when deleting non-existent note")
    void noteManager_deleteNonExistent_returnsFalse() {
        assertFalse(manager.deleteNote(999));
    }

    @Test
    @DisplayName("SaveNotesCommand should save notes to file")
    void saveNotesCommand_savesToFile() throws Exception {
        manager.addNote("Note 1", "Content 1");
        manager.addNote("Note 2", "Content 2");

        var command = new SaveNotesCommand(manager, TEST_FILE);
        command.execute();

        assertTrue(Files.exists(Path.of(TEST_FILE)));
    }

    @Test
    @DisplayName("LoadNotesCommand should load notes from file")
    void loadNotesCommand_loadsFromFile() throws Exception {
        manager.addNote("Note 1", "Content 1");
        manager.addNote("Note 2", "Content 2");
        manager.saveNotes(TEST_FILE);

        var newManager = new NoteManager();
        var command = new LoadNotesCommand(newManager, TEST_FILE);
        command.execute();

        assertEquals(2, newManager.listNotes().size());
    }

    @Test
    @DisplayName("NoteManager should preserve note IDs after save and load")
    void noteManager_saveAndLoad_preservesIds() throws Exception {
        var note1 = manager.addNote("Note 1", "Content 1");
        var note2 = manager.addNote("Note 2", "Content 2");
        manager.saveNotes(TEST_FILE);

        var newManager = new NoteManager();
        newManager.loadNotes(TEST_FILE);
        var loadedNotes = newManager.listNotes();

        assertTrue(loadedNotes.stream().anyMatch(n -> n.id() == note1.id()));
        assertTrue(loadedNotes.stream().anyMatch(n -> n.id() == note2.id()));
    }

    @Test
    @DisplayName("LoadNotesCommand should handle non-existent file")
    void loadNotesCommand_nonExistentFile_doesNotThrow() throws Exception {
        var command = new LoadNotesCommand(manager, "non_existent.txt");
        assertDoesNotThrow(() -> command.execute());
    }
}

File: kotlin/note-taking/src/test/kotlin/com/rbleggi/notetaking/NoteManagerTest.kt
package com.rbleggi.notetaking

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import java.io.File

class NoteManagerTest {

    private lateinit var manager: NoteManager

    @BeforeEach
    fun setup() {
        manager = NoteManager()
    }

    @Test
    fun `add note should create a new note`() {
        val note = manager.addNote("Test Note", "Test Content")
        assertEquals(1, note.id)
        assertEquals("Test Note", note.title)
        assertEquals("Test Content", note.content)
    }

    @Test
    fun `add multiple notes should increment id`() {
        val note1 = manager.addNote("Note 1", "Content 1")
        val note2 = manager.addNote("Note 2", "Content 2")
        assertEquals(1, note1.id)
        assertEquals(2, note2.id)
    }

    @Test
    fun `edit note should update title and content`() {
        manager.addNote("Original", "Original Content")
        val updated = manager.editNote(1, "Updated", "Updated Content")
        assertNotNull(updated)
        assertEquals("Updated", updated?.title)
        assertEquals("Updated Content", updated?.content)
    }

    @Test
    fun `edit note with only title should keep original content`() {
        manager.addNote("Original", "Original Content")
        val updated = manager.editNote(1, newTitle = "Updated")
        assertNotNull(updated)
        assertEquals("Updated", updated?.title)
        assertEquals("Original Content", updated?.content)
    }

    @Test
    fun `edit note with only content should keep original title`() {
        manager.addNote("Original", "Original Content")
        val updated = manager.editNote(1, newContent = "Updated Content")
        assertNotNull(updated)
        assertEquals("Original", updated?.title)
        assertEquals("Updated Content", updated?.content)
    }

    @Test
    fun `edit non-existent note should return null`() {
        val updated = manager.editNote(999, "Updated", "Updated Content")
        assertNull(updated)
    }

    @Test
    fun `delete note should remove it from manager`() {
        manager.addNote("Test Note", "Test Content")
        val deleted = manager.deleteNote(1)
        assertTrue(deleted)
        assertEquals(0, manager.listNotes().size)
    }

    @Test
    fun `delete non-existent note should return false`() {
        val deleted = manager.deleteNote(999)
        assertFalse(deleted)
    }

    @Test
    fun `list notes should return all notes`() {
        manager.addNote("Note 1", "Content 1")
        manager.addNote("Note 2", "Content 2")
        manager.addNote("Note 3", "Content 3")
        val notes = manager.listNotes()
        assertEquals(3, notes.size)
    }

    @Test
    fun `save and load notes should persist data`() {
        val testFile = "test-notes.txt"
        manager.addNote("Note 1", "Content 1")
        manager.addNote("Note 2", "Content 2")
        manager.saveNotes(testFile)

        val newManager = NoteManager()
        newManager.loadNotes(testFile)
        val notes = newManager.listNotes()
        assertEquals(2, notes.size)
        assertEquals("Note 1", notes.find { it.id == 1 }?.title)
        assertEquals("Note 2", notes.find { it.id == 2 }?.title)

        File(testFile).delete()
    }
}
